doc/_extensions/vuln_table.py

```diff
@@ -0,0 +1,133 @@
+from pathlib import Path
+import requests
+import json
+import pickle
+
+AZ_CONN_STR_PUBLIC = ";".join(
+    (
+        "DefaultEndpointsProtocol=https",
+        "EndpointSuffix=core.windows.net",
+        "AccountName=ncsdocsa",
+    )
+)
+"""Azure connection string (public acces)."""
+
+AZ_CONTAINER = "ncs-doc-vuln"
+"""Azure container."""
+
+logger = logging.getLogger(__name__)
+
+class VulnTable(SphinxDirective):
+
+    def run(self):
+        if not hasattr(self.env, "vuln_cache"):
+            return [nodes.paragraph(text="No vulnerabilities found")]
+
+        cache = self.env.vuln_cache
+        keys = cache["schema"]
+        table = nodes.table()
+        tgroup = nodes.tgroup(cols=len(keys))
+        for _ in range(len(keys)):
+            colspec = nodes.colspec(colwidth=1)
+            tgroup.append(colspec)
+        table += tgroup
+
+        # table head row
+        thead = nodes.thead()
+        tgroup += thead
+        row = nodes.row()
+        for key in keys:
+            entry = nodes.entry()
+            entry += nodes.paragraph(text=key)
+            row += entry
+        thead.append(row)
+
+        # table body
+        rows = []
+        for vuln in cache["body"]:
+            row = nodes.row()
+            rows.append(row)
+
+            for key in keys:
+                entry = nodes.entry()
+                para = nodes.paragraph(text=vuln[key])
+                entry += para
+                row += entry
+
+        tbody = nodes.tbody()
+        tbody.extend(rows)
+        tgroup += tbody
+
+        return [table]
+
+def progress(response: PipelineResponse) -> None:
+    """Download progress callback."""
+
+    mb_tot = response.context["data_stream_total"] / (1024 * 1024)
+    mb_curr = response.context["download_stream_current"] / (1024 * 1024)
+    pct = int((mb_curr / mb_tot) * 100)
+    s = term_width_line(
+        bold("Downloading vulnerability files... ")
+        + f"{mb_curr:.2f} of {mb_tot:.2f} Mb [{pct} %]"
+    )
+    logger.info(s, nonl=pct < 100)
+
+def download_vuln_table(app: Sphinx) -> None:
+    """Download vulnerability table from cache"""
+
+    # Check if local cache exists
+    try:
+        with open(Path(app.doctreedir) / Path("vuln_cache.pickle"), "rb") as cache_file:
+            local_cache = pickle.load(cache_file)
+    except FileNotFoundError:
+        logger.info("No vulnerability cache found locally")
+        local_cache = None
+
+    # Check internet connection
+    try:
+        requests.get("https://ncsdocsa.blob.core.windows.net", timeout=3)
+    except (requests.ConnectionError, requests.Timeout):
+        logger.info("Could not retrieve vulnerability table online")
+        if local_cache:
+            logger.info("Using local vulnerability cache")
+            app.env.vuln_cache = local_cache
+        return
+
+    cc = ContainerClient.from_connection_string(AZ_CONN_STR_PUBLIC, AZ_CONTAINER)
+    remote_files = sorted(
+        [b for b in cc.list_blobs()],
```

Comment:
nit: `list_blobs()` seems to accept `timeout` argument in seconds, so maybe use that instead of previous check

Reply:
The `timeout` parameter seems to limit subrequests made after Azure receives the initial request, not the actual wait time for the client. Additionally, `socket_timeout` and `connection_timeout` does not appear to have any effect, neither when supplied to the `ContainerClient` nor in the following method calls. I have yet to find a suitable replacement for the preceding connectivity check.